python_modules/dagster/dagster/_config/pythonic_config/conversion_utils.py

```diff
@@ -59,44 +58,60 @@ class cached_property:
 
 
 # This is from https://github.com/dagster-io/dagster/pull/11470
-def _apply_defaults_to_schema_field(field: Field, additional_default_values: Any) -> Field:
-    # This work by validating the top-level config and then
-    # just setting it at that top-level field. Config fields
-    # can actually take nested values so we only need to set it
-    # at a single level
-
-    evr = validate_config(field.config_type, additional_default_values)
-
-    if not evr.success:
-        raise DagsterInvalidConfigError(
-            "Incorrect values passed to .configured",
-            evr.errors,
-            additional_default_values,
+def _apply_defaults_to_schema_field(old_field: Field, additional_default_values: Any) -> Field:
+    """Given a config Field and a set of default values (usually a dictionary or raw default value),
+    return a new Field with the default values applied to it (and recursively to any sub-fields).
+    """
+    # If the field has subfields and the default value is a dictionary, iterate
+    # over the subfields and apply the defaults to them.
+    if isinstance(old_field.config_type, _ConfigHasFields) and isinstance(
+        additional_default_values, dict
+    ):
+        updated_sub_fields = {
+            k: _apply_defaults_to_schema_field(
+                sub_field, additional_default_values.get(k, FIELD_NO_DEFAULT_PROVIDED)
+            )
+            for k, sub_field in old_field.config_type.fields.items()
+        }
+
+        # We also apply a new default value to the field if all of its subfields have defaults
+        new_default = (
+            old_field.default_value if old_field.default_provided else FIELD_NO_DEFAULT_PROVIDED
         )
 
-    if field.default_provided:
-        # In the case where there is already a default config value
-        # we can apply "additional" defaults by actually invoking
-        # the config machinery. Meaning we pass the new_additional_default_values
-        # and then resolve the existing defaults over them. This preserves the default
-        # values that are not specified in new_additional_default_values and then
-        # applies the new value as the default value of the field in question.
-        defaults_processed_evr = resolve_defaults(field.config_type, additional_default_values)
-        check.invariant(
-            defaults_processed_evr.success,
-            "Since validation passed, this should always work.",
+        if all(
+            sub_field.default_provided or not sub_field.is_required
+            for sub_field in updated_sub_fields.values()
+        ) or (
+            old_field.config_type.kind == ConfigTypeKind.SELECTOR
+            and any(sub_field.default_provided for sub_field in updated_sub_fields.values())
+        ):
+            new_default = {
+                **additional_default_values,
+                **{k: v.default_value for k, v in updated_sub_fields.items() if v.default_provided},
+            }
```

Comment:
and then composing the generation of `new_defaults` into its own function might make code cleanar